server/src-test/Data/NonNegativeIntSpec.hs

```diff
@@ -0,0 +1,24 @@
+module Data.NonNegativeIntSpec (spec) where
+-- | basic tests on NonNegativeIntType
+
+import           Data.Aeson              (ToJSON (toJSON), Value (Number), decode)
+import           Hasura.RQL.Types.Common (NonNegativeInt (..))
+import           Prelude
+import           Test.Hspec              (Spec, describe, it, shouldBe, shouldNotBe)
+
+spec :: Spec
+spec = do
+  nonNegIntSpec
+
+nonNegIntSpec :: Spec
+nonNegIntSpec =
+    describe "non negative integer type" $ do
+        it "only validates non negative integers" $ do
+            NonNegativeInt 23 `shouldBe` 23
+            NonNegativeInt (-23) `shouldNotBe` (-23)
+
+        it "JSON accepts only non negative integers" $ do
+            toJSON (23 :: NonNegativeInt) `shouldBe` Number 23
+            toJSON (-23 :: NonNegativeInt) `shouldNotBe` Number (-23)
```

Comment:
- I can think of why L18 should be failing, there's nothing stopping me from setting that value.
 - I don't understand full well why L22 should be failing.

Reply:
IIUC L22 fails because you don't have a `ToJSON` instance which will reject from creating a JSON object if the value is negative. 

You have only created the `FromJSON` instance and that's why the `decode` tests below work.

Reply:
Line 18 and 22 fail because of the `NonNegativeInt` constructor. The `NonNegativeInt` is just a type, we need to control what value we construct, hence the check in the `FromJSON` instance i.e. if the number is negative we throw an error before constructing the `NonNegativeInt` value.